lib/hooks/queries/usePortfolioPositions.ts

```diff
@@ -0,0 +1,580 @@
+            return pool.marketId === getMarketIdOf(assetId);
+          }
+        });
+
+        if (!pool) return null;
+
+        const assetIds = pool.weights
+          .map((w) => fromCompositeIndexerAssetId(w.assetId).unwrap())
+          .filter(IOMarketOutcomeAssetId.is.bind(IOMarketOutcomeAssetId));
+
+        return assetIds.map((assetId) => ({
+          assetId,
+          account: poolAccountIds[pool.poolId],
+        }));
+      })
+      .filter(isNotNull) ?? [];
+
+  const poolAssetBalances = useAccountAssetBalances(poolAssetBalancesFilter);
+
+  const poolAssetBalances24HoursAgo = useAccountAssetBalances(
+    poolAssetBalancesFilter,
+    block24HoursAgo,
+    { enabled: Boolean(now?.block) },
+  );
+
+  const userAssetBalances = useAccountAssetBalances(
+    rawPositions.data?.map((position) => ({
+      assetId: fromCompositeIndexerAssetId(position.assetId).unwrap(),
+      account: address,
+    })) ?? [],
+  );
+
+  const positions = useMemo<Position[] | null>(() => {
+    let positionsData: Position[] = [];
+
+    if (!rawPositions.data || !pools.data || !markets.data) {
+      return null;
+    }
+
+    const calculatePrice = (
+      pool: IndexedPool<Context>,
+      assetId: AssetId,
+      poolsZtgBalances: PoolZtgBalanceLookup,
+      poolAssetBalances: UseAccountAssetBalances,
+    ): null | Decimal => {
+      const poolZtgBalance = poolsZtgBalances[pool.poolId]?.free.toNumber();
+
+      if (typeof poolZtgBalance === "undefined") {
+        return null;
+      }
+
+      const poolAssetBalance = poolAssetBalances.get(
+        poolAccountIds[pool.poolId],
+        assetId,
+      )?.data.balance;
+
+      if (!poolAssetBalance || isNA(poolAssetBalance)) return null;
+
+      const ztgWeight = new Decimal(pool.totalWeight).div(2);
+      const assetWeight = getAssetWeight(pool, assetId).unwrap();
+
+      return calcSpotPrice(
+        poolZtgBalance,
+        ztgWeight,
+        poolAssetBalance.free.toNumber(),
+        assetWeight,
+        0,
+      );
+    };
+
+    if (!rawPositions.data) return null;
+
+    let stillLoading = false;
+
+    for (const position of rawPositions.data) {
+      const assetId = fromCompositeIndexerAssetId(position.assetId).unwrap();
+
+      let pool: IndexedPool<Context>;
+      let marketId: number;
+      let market: FullMarketFragment;
+
+      if (IOZtgAssetId.is(assetId)) {
+        continue;
+      }
+
+      if (IOPoolShareAssetId.is(assetId)) {
+        pool = pools.data.find((pool) => pool.poolId === assetId.PoolShare);
+        marketId = pool.marketId;
+        market = markets.data?.find((m) => m.marketId === marketId);
+      }
+
+      if (IOMarketOutcomeAssetId.is(assetId)) {
+        marketId = getMarketIdOf(assetId);
+        market = markets.data?.find((m) => m.marketId === marketId);
+        pool = pools.data.find((pool) => pool.marketId === marketId);
+      }
+
+      if (!market || !pool) {
+        stillLoading = true;
```

Comment:
Currently it has to be like this because of some of the query hooks architecture, they dont return the query itself just data. I agree that this should be fixed, but in a different pr. It works quite well as it is though so I wont fix this on this pr.